### PR TITLE
Protocol params sanity checks

### LINKED ISSUE #512: Sanity checks for protocol parameters & derive parameters as much as possible
It would be great to have sanity checks for protocol parameters in place that evaluate whether a certain combination of parameters makes sense. This could, for example happen, when constructing the protocol parameters and the function that does all these checks could serve as a documentation going forward as to which parameters are related and how. 

The functionality should have different levels of escalation:
- for sensible parameters for a live network: this should warn, so that it's still possible to overwrite them e.g. in unit or end-to-end tests
- impossible combinations of parameters: this should panic

### PR DESCRIPTION
**Pull request by cyberphysic4l**
Add sanity checks to protocol parameters where we have certain constraints e.g. MaxCommittableAge >= MinCommittableAge

closes https://github.com/iotaledger/iota-core/issues/512

### CODE REVIEW

**Comment by karimodm** [comment]
We are recently having discussions about deriving as many protocol parameters as possible to partially avoid these sanity checks. Would this be the right PR to start deriving some?

**Comment by cyberphysic4l** [comment]
> We are recently having discussions about deriving as many protocol parameters as possible to partially avoid these sanity checks. Would this be the right PR to start deriving some?

Yes, I think that would be a good task to combine with deriving some. Maybe let's start making a list of what we might like to make derived in the issue? https://github.com/iotaledger/iota-core/issues/512

**Comment by oliviasaa** [comment]
can we add these checks as well?
https://github.com/iotaledger/iota-core/issues/450

**Comment by oliviasaa** [comment]
what is missing is:
- decayFactorEpochsSum*generationRate must use at most 32 bits
- 21 * S * gamma * d/(beta*Delta) must use at most bitsCount bits

where 
- S is the token supply, 
- gamma is the generation rate per slot, meaning generationRate*2^(-generationRateExp)
- d is the number of slots per epoch
- beta is the decay factor per year (1/3)
- Delta is the epoch duration in years

This is something that cannot be "derived"; those are consistency checks to prevent overflowing in some functions.

**Comment by oliviasaa** [comment]
I'm having a problem testing this repo locally and I don't want to break anything. I created another branch so you can see what I did.

File: api_v3_protocol_parameters.go [hunk 1]
```diff
@@ -310,17 +349,28 @@ func WithManaOptions(bitsCount uint8, generationRate uint8, generationRateExpone
...
 
-func WithTimeProviderOptions(genesisSlot SlotIndex, genesisTimestamp int64, slotDurationInSeconds uint8, slotsPerEpochExponent uint8) options.Option[V3ProtocolParameters] {
+func WithTimeOptions(genesisSlot SlotIndex, genesisTimestamp int64, slotDurationInSeconds uint8, slotsPerEpochExponent uint8, livenessThresholdLowerBoundInSeconds uint16, livenessThresholdUpperBoundInSeconds uint16, minCommittableAge SlotIndex, maxCommittableAge SlotIndex, epochNearingThreshold SlotIndex) options.Option[V3ProtocolParameters] {
+	if livenessThresholdLowerBoundInSeconds > livenessThresholdUpperBoundInSeconds {
+		panic("livenessThresholdLowerBoundInSeconds must be smaller than livenessThresholdUpperBoundInSeconds")
+	}
+	if SlotIndex(livenessThresholdUpperBoundInSeconds) > minCommittableAge*SlotIndex(slotDurationInSeconds) {
+		panic("livenessThresholdUpperBoundInSeconds * slotDurationInSeconds must be smaller than minCommittableAge")
+	}
+	if minCommittableAge > maxCommittableAge {
```
Comment (bingyanglin) [inline api_v3_protocol_parameters.go]: >=

File: api_v3_protocol_parameters.go [hunk 2]
```diff
@@ -310,17 +349,28 @@ func WithManaOptions(bitsCount uint8, generationRate uint8, generationRateExpone
 	}
 }
 
-func WithTimeProviderOptions(genesisSlot SlotIndex, genesisTimestamp int64, slotDurationInSeconds uint8, slotsPerEpochExponent uint8) options.Option[V3ProtocolParameters] {
+func WithTimeOptions(genesisSlot SlotIndex, genesisTimestamp int64, slotDurationInSeconds uint8, slotsPerEpochExponent uint8, livenessThresholdLowerBoundInSeconds uint16, livenessThresholdUpperBoundInSeconds uint16, minCommittableAge SlotIndex, maxCommittableAge SlotIndex, epochNearingThreshold SlotIndex) options.Option[V3ProtocolParameters] {
+	if livenessThresholdLowerBoundInSeconds > livenessThresholdUpperBoundInSeconds {
+		panic("livenessThresholdLowerBoundInSeconds must be smaller than livenessThresholdUpperBoundInSeconds")
+	}
+	if SlotIndex(livenessThresholdUpperBoundInSeconds) > minCommittableAge*SlotIndex(slotDurationInSeconds) {
```
Comment (bingyanglin) [inline api_v3_protocol_parameters.go]: >=

**Comment by oliviasaa** [comment]
TO DO:

- [x] derive bootstrapping duration from beta
- [x] review sanity checks for decay factor sum and other mana related parameters

File: mana_decay_provider.go [hunk 3]
```diff
@@ -244,35 +237,46 @@ func (p *ManaDecayProvider) GenerateManaAndDecayBySlots(amount BaseToken, creati
 		if err != nil {
 			return 0, ierrors.Wrap(err, "failed to decay generated mana in the first epoch")
 		}
+		/*
+			manaGeneratedWholeEpoch, err := p.generateMana(amount, 1<<p.slotsPerEpochExponent)
```
Comment (muXxer) [inline mana_decay_provider.go]: what is that here?

Comment (cyberphysic4l) [reply mana_decay_provider.go]: Don't know, some workings from Olivia... Deleted.

**Review by muXxer — APPROVED** [review]

**Pull request merged**